### PR TITLE
rbd_mirror/Replayer: make Replayer::set_sources() thread-safe

### PR DESCRIPTION
**Pull request by sahithi-rv**
Signed-off-by: Sahithi R V <sahithi.rv1@gmail.com>

### CODE REVIEW

**Comment by sahithi-rv** [comment]
Here, only m_images is locked because it is the only shared data structure. 
I would like to know, what is the condition, which the m_cond is waiting on ?(m_cond.WaitInterval())

**Comment by jdurgin** [comment]
m_images needs to be protected while it's being read too, so it was simpler to take it for the whole method. To make creating and removing the ImageReplayers async, we need to wait for it to be implemented in #7614 and then make the constructor and destructor not do any blocking work, and make asynchronous start and stop methods do the work that would be blocking, or something along those lines.

m_cond is passed by reference to the PoolWatcher, which would trigger it whenever it detects new images. Right now it doesn't do much, since it's just polling. It's intended to be helpful when we have watch/notify notifications of mirrored image creation/deletion, so we can respond to those events rather than polling for changes.

**Comment by dillaman** [comment]
@sahithi-rv Can assign the associated tracker ticket to yourself?  Thx.

**Comment by dillaman** [comment]
I believe this task will be dependent upon task 14530.  The end goal that we want to achieve is to avoid any blocking when opening and closing images and their associated journals.  Blocking can occur because of locks, but in the distributed world of Ceph, it also occurs when we send synchronous requests to the OSDs.  Since we might have thousands of images to synchronize, any blocking while processing one image would prevent forward progress on the remainder of images handled by the process.

**Comment by sahithi-rv** [comment]
@dillaman , I am trying to understand issue 14530 and work on it.

**Comment by sahithi-rv** [comment]
@jdurgin , thanks. I will think better and work on it.

**Comment by sahithi-rv** [comment]
@jdurgin Please check my latest commit.

**Comment by sahithi-rv** [comment]
@dillaman , I am having difficulty in understanding how is issue 14530 related to rbd_mirror. You have previously mentioned, JournalTrimmer and JournalMetadata might send messages to OSDs and that has to be done asynchronously.

**Comment by dillaman** [comment]
@sahithi-rv Yup -- ImageReplayer uses the journal code, so in order to asynchronously open an close image replayers, we must also be able to asynchronously open and close the journal.

File: src/tools/rbd_mirror/Replayer.cc [hunk 1]
```diff
@@ -93,21 +93,29 @@ int Replayer::init()
 void Replayer::run()
 {
   while (!m_stopping.read()) {
-    Mutex::Locker l(m_lock);
     set_sources(m_pool_watcher->get_images());
-    m_cond.WaitInterval(g_ceph_context, m_lock, seconds(30));
+    //m_cond.WaitInterval(g_ceph_context, m_lock, seconds(30));
```
Comment (jdurgin) [inline src/tools/rbd_mirror/Replayer.cc]: to summarize our irc discussion: this m_cond.WaitInterval() call is still needed here, so we don't call set_sources() continuously

Also, this commit is making set_sources() thread-safe, which is different from #14673, so the commit message should be updated to remove #14673

Comment (jdurgin) [inline src/tools/rbd_mirror/Replayer.cc]: you might use something like "rbd_mirror: make Replayer::set_sources() thread-safe" as a title for the commit instead

File: src/tools/rbd_mirror/Replayer.cc [hunk 2]
```diff
@@ -101,13 +100,22 @@ void Replayer::run()
 
 void Replayer::set_sources(const map<int64_t, set<string> > &images)
 {
-  assert(m_lock.is_locked());
+  std::map<int64_t, std::map<std::string,
+                 std::unique_ptr<ImageReplayer> > > m_images_copy;
```
Comment (dillaman) [inline src/tools/rbd_mirror/Replayer.cc]: Minor: the ```m_``` prefix should only be used for class-level variables.

**Comment by dillaman** [comment]
@sahithi-rv Can you rebase this PR?  I am seeing the following compile error:

> error: tools/rbd_mirror/Replayer.cc:103:83: use of deleted function 'rbd::mirror::ImageReplayer::ImageReplayer(const rbd::mirror::ImageReplayer&)'

**Comment by sahithi-rv** [comment]
rebased.

**Comment by dillaman** [comment]
@sahithi-rv Still seeing the same compile error

**Comment by sahithi-rv** [comment]
@dillaman I will work on it , and try to fix it.

**Comment by sahithi-rv** [comment]
The error here is because of https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/ImageReplayer.h#L70. So, I understand that this is because of making a constant instance of ImageReplayer. But because, unique_ptr is used to point to ImageReplayer, I don't find any other way of making a copy of m_images. Any suggestions? Can we do anything about L70 of ImageReplayer.h

**Comment by dillaman** [comment]
I am probably missing something, but is ```Replayer::set_sources``` the only code that references ```m_images```?  If so, it already appears to be thread-safe since it is only being invoked by its own thread.

**Comment by sahithi-rv** [comment]
@dillaman Yes, I think you are right.

**Pull request closed without merging**